### ISSUE TITLE
Can't compress DOS 32-bit EXE (Coff) files generated by Freepascal 3.0.x

### ISSUE THREAD

**Issue by Laaca** [upx/upx#45]
## What's the problem (or question)?

No .EXE file compiled by Freepascal 3.0.x (at least the DOS executables) can't be compressed using UPX. Any attempt to compress it leads into error "AlreadyPackedException". 
The Freepascal is self compiling so the best way how to test it is just download this DOS instalator executable: ftp://gd.tuwien.ac.at/languages/pascal/fpc/dist/3.0.0/i386-go32v2/separate/install.exe

## What should have happened?

This problem was not in Freepascal 2.x.y or Freepascal 1.x.y.
It seems that the FPC version is stored somewhere in EXE header and somehow interferes with some UPX flag.

## How can we reproduce the issue?

1.
2.
3.
4.

## Please tell us details about your environment.

Should be 100% reproducible however I use FreeDOS or Windows 98.

**Comment by ml1050** [comment]
install.exe has a little bit different header layout compared to "older" djgpp programs, and UPX currently does not support this. Printing AlreadyPackedException is incorrect, but this is what we have now.

Adding support for this feature is not too hard, but I touched that code 10 years ago, which means I could easily break something else while implementing this feature. I do not promise any deadline for this.

**Comment by markus-oberhumer** [comment]
As a first step, it would be nice if we can report a better error message.

**Comment by ssokolow** [comment]
> No .EXE file compiled by Freepascal 3.0.x (at least the DOS executables)

Just the DOS ones. I have a project I'm just starting that is targeted at DOS (essentially, an install wizard built as a self-extractor stub) but can also build under the Linux for rapid prototyping and the Linux builds compress and function just fine.

(And it gets a ~37% ratio on Linux when using `--ultra-brute`, so it's a shame that UPX can't be used on the GO32v2 build.)

**Comment by Laaca** [comment]
Please remove the stale sign.

**Comment by jreiser** [comment]
@Laaca  Please update the URL of the testcase.  When I tried just now, the host could not be found (and http: also failed.)
```
$ curl -O ftp://gd.tuwien.ac.at/languages/pascal/fpc/dist/3.0.0/i386-go32v2/separate/install.exe
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (6) Could not resolve host: gd.tuwien.ac.at
```

**Comment by jreiser** [comment]
Perhaps https://sourceforge.net/projects/freepascal/files/Win32/3.0.0/fpc-3.0.0.i386-win32.exe/download ?

**Comment by jreiser** [comment]
And the failure now is
```
Breakpoint 1, throwInternalError (msg=0x7fffffffc8a0 "ElfLinker output_capacity 0 exceeded; now 0 want 20 for Dgbgehggbhagjdddccogegmgm\n") at except.cpp:124
```

**Comment by Laaca** [comment]
No, the link you provided points to Win32 installer and it si a Windows 32-bit PE exacutable.
(ironicaly, it also failures)

The URL for the DOS installer (which is a 32-bit DOS Coff application) is f.e. here:
https://sourceforge.net/projects/freepascal/files/DOS_Go32v2%20Single%20Pieces/3.2.2/install.exe/download

**Comment by jreiser** [comment]
Please try commit eac82ebfaf263ee0ed017874c4f2f7283ebbd73e on `devel` branch, which compresses install.exe without complaint.  I am not equipped to test execution.

**Comment by Laaca** [comment]
OK, I will try to set up a GNU C environment and then to build a binary from your sources :-)

**Comment by jreiser** [comment]
@Laaca Click on the green check-mark next to "d46e824" above, to see the auto-builds.

**Comment by Laaca** [comment]
OK, I tried the autobuild for Windows32. The DOS/Coff (brewed by Freepascal 3.0.4) files are compressed without any complaints. So far good. But the resulting compressed binaries do not work more. They end with General protection fault.
To be precide - the very minimal "Hello world!" program after compression works but the slightly modified version (Hello world from subdirectory ....) does not.
Look here http://www.laaca.borec.cz/hello1.zip and here [http://www.laaca.borec.cz/hello1.zip](url)
However after decompression (UPX -d) it works again.
The same applies to our test case with INSTALL.EXE
[https://sourceforge.net/projects/freepascal/files/DOS_Go32v2%20Single%20Pieces/3.2.2/install.exe/download](url)

**Comment by ssokolow** [comment]
Y'know, this stale bot gives a strong impression that you care more about chasing a low open bugs metric than actually fixing things.

I'm reminded of the similar "want to look like they care about users but unwilling to put in the honest work" aura of JWZ's [CADT Model](https://www.jwz.org/doc/cadt.html).

**Comment by markus-oberhumer** [comment]
This is our spare-time hobby project, supporting 44 different executable formats.

So yes, in this Open Source project it is actually the responsibility of our **users** to provide helpful bug reports and motivation to fix things.

If nobody cares about some obscure bug for more than 90 days, why should we?

https://github.com/upx/upx/blob/master/.github/workflows/close-stale-issues.yml :

```
# Automatically closing issues is far from perfect, but then we only have
# limited resources and this approach favors issues that people
# actually care about.
#
# also see:
#   "Should I Stale or Should I Close? An Analysis of a Bot that Closes
#    Abandoned Issues and Pull Request".
#   https://igorwiese.com/images/papers/Paper_BotSE_19.pdf

```

**Comment by ssokolow** [comment]
Maybe add a sentence to the bot message that summarizes and hyperlinks to that rationale then. As-is, it's terrible PR.

**Comment by Laaca** [comment]
I the original reporter of this bug and I am happy that this bug can be closed as resolved.
UPX 4.0 can (finally!!!) compress the Freepascal 3.x.y files.
